language/types_values_variables.md

```diff
@@ -1463,6 +1476,100 @@ $a = Binary('YWJj')
+except `scheme`. The following diagram shows how the parts are mapped:
+
+Hierarchical `URI`:
+~~~
+  abc://username:password@example.com:123/path/data?key=value&key2=value2#fragid1
+  └┬┘   └───────┬───────┘ └────┬────┘ └┬┘└────┬───┘ └─────────┬─────────┘ └──┬──┘
+scheme      userinfo         host    port   path            query         fragment
+~~~
+Opaque `URI`:
+~~~
+  urn:example:mammal:monotreme:echidna
+  └┬┘ └──────────────┬───────────────┘
+scheme             opaque
+~~~
+
+The type can be constrained to match instances on a fine grained level using the `Hash` where each entry except the `port`
+can be constrained using:
+
+* A `String` that represents an exact match.
+* A `Regexp` that represents an pattern match
+* A `Type[Pattern]` that may contain one or more regexps
+* A `Type[Enum]` where at least one of the included strings must match
+* A `Type[NotUndef]` representing that the entry must be present 
+* An `undef` or `Type[Undef]` representing that the entry must not be present
+
+The `port` entry can be constrained using:
+
+* A `Type[Integer]` that represents the range of valid port numbers
+* A `Type[NotUndef]` representing that the entry must be present 
+* An `undef` or `Type[Undef]` representing that the entry must not be present
+
+When `V` is a string, it will be parsed into a `Hash` where each contained part will be a `String` except `port` which will be
+an `Integer`.
+
+***Using an URI instance***
+
+An `URI` instance can be created from a `String` using valid syntax or a `Hash` using the valid part names.
+
+Individual parts of an `URI` are available as instance attrbiutes and an `URI` can be merged with another `URI` or a `String`
+using the `+` operator.
+
+**Examples:**
+
+Creating a URI from a string
+
+```puppet
+$u = URI('http://www.example.com')
+```
+
+Creating a URI from a hash
+
+```puppet
+$u = URI(scheme => 'http', host => 'www.example.com', path => '/a/b')
+```
+
+Accessing parts as instance attributes:
+
+```puppet
+$u = URI('http://bob:pwd@www.example.com:23/a/b?x=2#frag')
+notice($u.scheme)   # will notice the string 'http'
+notice($u.userinfo) # will notice the string 'bob:pwd'
+notice($u.host)     # will notice the string 'www.example.com'
+notice($u.port)     # will notice the string '23' (from integer value)
+notice($u.path)     # will notice the string '/a/b'
+notice($u.query)    # will notice the string 'x=2'
+notice($u.fragment) # will notice the string 'frag'
+
+$o = URI('urn:a:b:c')
+notice($o.scheme)   # will notice the string 'urn'
+notice($o.opaque)   # will notice the string 'a:b:c'
+```
+
+Merging URIs:
+
+```puppet
+$d = URI('http://example.com/a/b/')
+$ap = URI('/c/d')
+notice($d + $ap) # notices 'http://example.com/c/d'
+
+$rp = URI('c/d')
+notice($d + $rp) # notices 'http://example.com/a/b/c/d'
+
+$f = URI('http://example.com/a/b')
+notice($f + $rp) # notices 'http://example.com/a/c/d'
+```
+
```

Comment:
Would like to see a couple of examples with matching types - esp using constraints - for instance restricting a URI to not accept fragments / query, only http and https and that path must be included are useful I think.